plutus-benchmark/marlowe/test/Lib.hs

```diff
@@ -0,0 +1,169 @@
+    reportDifference
+    (createDirectoriesAndWriteFile goldenPath)
+    (removeFile outputPath)
+ where
+  reportDifference :: Text -> Text -> IO (Maybe String)
+  reportDifference expected actual = do
+    let parse = traverse parseLine . Text.lines
+    expectedResults <- parse expected
+    actualResults <- parse actual
+    if expectedResults == actualResults
+      then pure Nothing
+      else do
+        let (cpu0, mem0, size0) = aggregateResults expectedResults
+            (cpu1, mem1, size1) = aggregateResults actualResults
+
+        pure . Just $
+          "Actual execution budgets and sizes differ from expected ones:\n"
+            <> Tabular.render
+              id
+              id
+              Text.unpack
+              ( Table
+                  ( Group
+                      NoLine
+                      [ Header "Baseline (average)"
+                      , Header "Actual (average)"
+                      , Header "Delta"
+                      ]
+                  )
+                  ( Group
+                      SingleLine
+                      [ Header "CPU units"
+                      , Header "Memory units"
+                      , Header "Size"
+                      ]
+                  )
+                  [ [formatUnits cpu0, formatUnits mem0, formatUnits size0]
+                  , [formatUnits cpu1, formatUnits mem1, formatUnits size1]
+                  ,
+                    [ formatDelta cpu1 cpu0
+                    , formatDelta mem1 mem0
+                    , formatDelta size1 size0
+                    ]
+                  ]
+              )
+
+aggregateResults
+  :: [(ExCPU, ExMemory, UPLC.Size)]
+  -> (Integer, Integer, Integer)
+aggregateResults results =
+  ( average (map (\(ExCPU i) -> fromSatInt i) cpus)
+  , average (map (\(ExMemory m) -> fromSatInt m) mems)
+  , average (map UPLC.unSize sizes)
+  )
+ where
+  (cpus, mems, sizes) = unzip3 results
+  average xs
+    | null xs = 0
+    | otherwise = sum xs `div` fromIntegral (length xs)
+
+formatDelta :: Integer -> Integer -> Text
+formatDelta a b = F.sformat (F.fixed 2 F.% "%") (delta b a)
+ where
+  delta :: Integer -> Integer -> Double
+  delta x y
+    | x > 0 = (fromIntegral (y - x) * 100) / fromIntegral x
+    | otherwise = 0
+
+formatUnits :: Integer -> Text
+formatUnits = F.sformat (F.groupInt 3 ' ')
+
+-- | Like 'writeFile', but also create parent dirs if they are missing.
+createDirectoriesAndWriteFile :: FilePath -> Text -> IO ()
+createDirectoriesAndWriteFile path bs = do
+  let dir = Path.takeDirectory path
+  createDirectoryIfMissing
+    True -- create parents too
+    dir
+  Text.IO.writeFile path bs
+
+parseLine :: Text -> IO (ExCPU, ExMemory, UPLC.Size)
+parseLine ln =
+  case Text.splitOn "\t" ln of
+    [cpu, mem, size] ->
+      (,,)
+        <$> readCpu cpu
+        <*> readMem mem
+        <*> readSize size
+    _ -> throwIO . userError $ "Can't parse line: " <> show ln
+ where
+  readCpu :: Text -> IO ExCPU
```

Comment:
That's quite a lot of work just to read three numbers, but I don't know if refactoring it would make it any shorter.